Try alternate style to indicate selected blocks

jasmussen opened the pull request:
## Description

This PR takes a first stab at some of the changes discussed/mocked up in #2983.

Basically it removes both the hover and selected styles for blocks, in favor of little floating mover/settings blobs on the side, and some subtler selection styles for a few blocks (separator, more, button). 

Additionally, it changes the visual style of multi-selected blocks. Given incoming keyboard-multi-select features, this changes it so a multi selected block has the same color as highlighted text does. This uses a kind of magical CSS value, `Highlight`, which is documented in https://www.sitepoint.com/css-system-styles/. Video:

![behavior](https://user-images.githubusercontent.com/1204802/31773494-3ddea31e-b4e3-11e7-96f3-eb4c7f86b6a8.gif)

This property is quite magical, and actually changes along with whatever choice you make at the operating system level:

![itshappening](https://user-images.githubusercontent.com/1204802/31773759-07cc1f26-b4e4-11e7-97ab-00454ef314df.gif)

It's a CSS 2.1 property and I've tested this color and found it working all the way back to IE9 (didn't test further). 

I'm marking this as "In Progress", as I'd like both some code reviews, as well as some _experience_ reviews. How does this feel? What feels good? What feels not so good? What else needs to be done in this branch? Any experience regressions?

Keep in mind that this is only one aspect of the visual refresh, you can see a list of other complimentary tasks here: https://github.com/WordPress/gutenberg/issues/2983#issuecomment-337841535

## How Has This Been Tested?

Tested the color effect in Safari, Firefox, Chrome both on Mac and Windows, as well as Edge and IE9-11 on Windows. Not that we need 9 support, I was just impressed that the CSS property worked all the way back for that one.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

Commit:
First pass at alternate block style.

Commit:
First pass at alternate block style.

Commit:
Polish the hovered mover and settings.

Commit:
Polish the hovered mover and settings.

Commit:
Try a subtle "text focus" color for some blocks.

youknowriad commented on editor/modes/visual-editor/style.scss:
```diff
@@ -77,24 +74,35 @@
-		outline: 1px solid $light-gray-500;
-		transition: 0.2s outline;
+	// simpler style for a block that has cursor focus (but hasn't been selected)
+	&.is-selected .editor-block-mover:before,
```
We may want to unify the className used for the `hover` and `selected`, maybe something like `is-ui-visible`

youknowriad reviewed (commented):
My experience with this:

- It feels and looks nice, 
- I don't feel the heavy blocks anymore. 
- It looks like it's one and unique content editable (even if it's not)
- Sometimes it hard to know what are the limits where the inspector/block settings will apply. I don't feel it's blocking but maybe just something we'll get used to over time.

afercia commented:
To my knowledge, system colors are deprecated in CSS3 and there's no guarantee they will work in the future.

https://www.w3.org/TR/css3-color/#css-system
> CSS2 system colors
> Deprecated.

> Note. The CSS2 System Color values have been deprecated in favor of the CSS3 UI ‘appearance’ property. If you want to emulate the look of a user interface related element or control, please use the ‘appearance’ property instead of attempting to mimic a user interface element through a combination of system colors.

mtias commented on editor/modes/visual-editor/style.scss:
```diff
@@ -284,6 +297,29 @@
 	}
 }
 
+// show a separate cursor/caret highlight style for when an item was selected by the mouse or keyboard (but isn't multi selected)
```
How would this scale? Is this something a block author should be able to opt into? What are the reasons for a block to be part of this group?

mtias commented:
Let's try this. I think we can get it for 1.5 and get some feedback. I'd like to test it before getting the docked toolbar PR merged.

afercia commented:
> Let's try this.

Sorry I can't understand why you want to merge an officially deprecated CSS feature.

jasmussen commented:
Thanks for the reviews.

> Sometimes it hard to know what are the limits where the inspector/block settings will apply. I don't feel it's blocking but maybe just something we'll get used to over time.

> How would this scale? Is this something a block author should be able to opt into? What are the reasons for a block to be part of this group?

Answering these together, because they are related. The answer is it doesn't scale well, and I'm still thinking about the best way to do this. 

The _idea_ is that when you have the cursor in _text_, the cursor itself is an indication of what is your focus. We need something _like that_, for when your "focus selection" is on something non text, like a separator — just to help you keep your place as to what's selected. 

There's also the thing that when we get across block keyboard selection with shift, the boundaries between paragraph blocks become less important. 

Whether _the block itself_ is important to show the boundaries of, remains to be seen, which is why I asked for your thoughts. I think if we can do some improvements around multi selection (see the recent keyboard selection discussion), perhaps clicking the ellipsis or mover also _selects_ the block (as in multi select indication), which will show the boundaries clearly.

Commit:
Add normal color as fallback for Highlight.

jasmussen commented:
> To my knowledge, system colors are deprecated in CSS3 and there's no guarantee they will work in the future.

I added a fallback color. If for whatever reason the `Highlight` color won't work, the background will be used instead. 

Alternatively we can change the color of the selection using `::selection`? But this seems like a better approach.

pento commented:
I rebased this, because it happened to be branched when `master` was going through an awkward buggy phase.

afercia commented:
> I added a fallback color. If for whatever reason the Highlight color won't work, the background will be used instead.

That's better 🙂 However, in my mind "deprecated" means "no use", especially in a new software. Note the Sitepoint post you linked to is from August 11, 2009.

jasmussen commented:
I understand the concern with using deprecated properties, but in this case we have fallback plans and alternatives, should they stop working. And while they are working, it seems like a nice usability win to use system colors.

Commit:
Restore the boundary for a selected block.

It's important for some blocks, if we are to look at removing this entirely, let's do that as a separate PR.

This does tweak the width to 1px though.

Commit:
Improve and polish the position of movers and settings.

Commit:
Fix issues with prompt and full-wide ellipsis.

jasmussen commented:
With the recent changes above, I believe this PR is ready for final review. As it won't go in 1.5, and folks are all busy getting that ship shape, I won't ping anyone, but anyone receiving these notifications, feel free to review when you're ready!

ellatrix reviewed (approved):
This looks good, after a rebase :)

Commit:
Fix rebase issues.

jasmussen commented:
Thanks Ella for the review and thumbs up. 

The rebase was a bit tricky — my own fault for making so many commits — but could I ask for your eyes on this another time just to verify it's okay?

mtias committed:
Count arrow keys for isTyping.

mtias commented:
Rebased to pull #2988.

mtias commented on editor/modes/visual-editor/block.js:
```diff
@@ -283,6 +283,7 @@ class VisualEditorBlock extends Component {
 			], this.props.order + 1 );
 		}
 		this.removeOrDeselect( event );
+		this.maybeStartTyping( event );
```
This is just to see how it feels—I like that arrow keys trigger "typing" mode now. But we'd want to exclude modifier keys, etc. cc @iseulde @mcsf

aduth replied:
Pushed 118f489ecace3f780a7bcf2ea8fe14d5ff2c24f6 to consolidate keydown handling, targeting arrow keys explicitly for triggering typing mode.

This reverts some previous refactoring that had split out `removeOrDeselect` to a separate function, but I think the switch is a bit easier to reason about consistently. Otherwise maybe some mapping of keys and conditions (`isTarget`) via something like [`cond`](https://lodash.com/docs/4.17.4#cond) but seems overkill and prone to a performance impact on this performance-critical code.

mtias replied:
Looks good to me.

aduth committed:
Block: Consolidate keydown event handling

mtias commented:
Let's 🚢  this.

aduth commented:
Noted today that the multi-select system highlight color is pretty extreme in Firefox (56.0.2, macOS 10.12.6)

![Firefox Multi-Select](https://user-images.githubusercontent.com/1779930/32612548-8b9e6c70-c536-11e7-9c7c-70596f56f874.png)

jasmussen commented:
That is pretty funky indeed. Is that the stable channel? What's the normal selection color?

afercia commented:
Firefox on Windows (IE 11 and Edge use the same color, Chrome a lighter one):

![screenshot 118](https://user-images.githubusercontent.com/1682452/32617660-d7657344-c575-11e7-892a-3fdd3a84ab4e.png)

aduth commented:
The normal selection color in Firefox is more reasonable, only the darker blue when multi-select takes effect.

![image](https://user-images.githubusercontent.com/1779930/32617734-c7ac7a38-c543-11e7-95a8-9f2ac7dc4863.png)

jasmussen commented:
I'll revisit. Thanks for the update.
